libcudacxx/include/cuda/std/__cccl/builtin.h

```diff
@@ -135,6 +147,32 @@
 #  define _CCCL_BUILTIN_EXPECT(...) __builtin_expect(__VA_ARGS__)
 #endif // _CCCL_CHECK_BUILTIN(builtin_expect)
 
+#if _CCCL_HAS_BUILTIN(__builtin_FILE) || defined(_CCCL_COMPILER_GCC) \
+  || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927)
+#  define _CCCL_BUILTIN_FILE() __builtin_FILE()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_FILE) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_FILE) vvv
+#  define _CCCL_BUILTIN_FILE() __FILE__
+#endif // !_CCCL_HAS_BUILTIN(__builtin_LINE)
+
+// NVCC below 11.3 cannot handle __builtin_FILE
+#if defined(_CCCL_CUDACC_BELOW_11_3)
+#  undef _CCCL_BUILTIN_FILE
+#  define _CCCL_BUILTIN_FILE() __FILE__
+#endif // _CCCL_CUDACC_BELOW_11_3
+
+#if _CCCL_HAS_BUILTIN(__builtin_FUNCTION) || defined(_CCCL_COMPILER_GCC) \
+  || (defined(_CCCL_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927)
+#  define _CCCL_BUILTIN_FUNCTION() __builtin_FUNCTION()
+#else // ^^^ _CCCL_HAS_BUILTIN(__builtin_FUNCTION) ^^^ / vvv !_CCCL_HAS_BUILTIN(__builtin_FUNCTION) vvv
+#  define _CCCL_BUILTIN_FUNCTION() "__builtin_FUNCTION is unsupported"
```

Comment:
C99 and C++11 added `__func__`, all compilers i know have `__FUNCTION__`.

Reply:
The issue is that both macros are only allowed within the body of a function, see https://godbolt.org/z/s3Wcv5T11

Reply:
what does `__builtin_FUNCTION()` do outside of a function? when would you want to use it like that?

Reply:
Its doing what is needed for source location. All those macros were effectively used for that